### PR TITLE
[ws client]: add function to detect if the background task is canceled.

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

### CODE REVIEW

File: ws-client/src/client.rs [hunk 1]
```diff
@@ -80,15 +81,17 @@ impl ErrorFromBack {
 /// It's possible that the background thread is terminated and this makes the client unusable.
 /// An error [`Error::RestartNeeded`] is returned if this happens and users has to manually
 /// handle dropping and restarting a new client.
-#[derive(Debug)]
+#[derive(Clone, Debug)]
```
Comment (niklasad1) [inline ws-client/src/client.rs:84]: we don't want this to be clone anymore for two reasons:

1. Uniform API with the `HTTP client`
2. Indicate to users that `Clone` means sharing resources such that users has to put in a proper such as in `Arc` or `Mutex` or whatever fits the particular usecase.

File: ws-client/src/client.rs [hunk 2]
```diff
@@ -157,6 +160,8 @@ impl WsClient {
 			.await
 			.map_err(|e| Error::TransportError(Box::new(e)))?;
 
+		let (shutdown_sender, shutdown_receiver) = async_std::channel::unbounded::<()>();
```
Comment (niklasad1) [inline ws-client/src/client.rs:163]: this is quite neat but no need for an unbounded channel, I think `Arc<AtomicBool>` would be simple enough for this usecase.

an alternative is to maybe put `error` in a `RwLock` then read it in `fn is_connected()`

File: ws-client/src/client.rs [hunk 3]
```diff
@@ -167,8 +172,19 @@ impl WsClient {
...
+			to_back,
+			request_timeout,
+			error: Arc::new(Mutex::new(ErrorFromBack::Unread(err_rx))),
+			shutdown_receiver,
+		})
+	}
+
+	/// True if the backend has not shutdown
+	pub async fn is_connected(&self) -> bool {
+		!self.shutdown_receiver.is_closed()
```
Comment (niklasad1) [inline ws-client/src/client.rs:187]: I suppose this would be called quite frequently w.r.t your use-case and using a Mutex/RwLock is not good idea.

**Review by niklasad1 — CHANGES_REQUESTED** [review]
Thanks for your PR.

We don't want the `Client` to implement `Clone` anymore, see my inline comments.

File: ws-client/src/client.rs [hunk 4]
```diff
@@ -167,8 +175,14 @@ impl WsClient {
 				max_concurrent_requests,
 			)
 			.await;
+			has_shutdown_sender.swap(true, Ordering::Relaxed);
```
Comment (niklasad1) [inline ws-client/src/client.rs:178]: why swap? the return value is ignored anyway?

I think `load` is fine here.

**Review by niklasad1 — APPROVED** [review]

File: ws-client/src/client.rs [hunk 5]
```diff
@@ -167,8 +175,14 @@ impl WsClient {
...
 			)
 			.await;
+			has_shutdown_sender.store(true, Ordering::Relaxed);
 		});
-		Ok(Self { to_back, request_timeout, error: Mutex::new(ErrorFromBack::Unread(err_rx)) })
+		Ok(Self { to_back, request_timeout, error: Mutex::new(ErrorFromBack::Unread(err_rx)), has_shutdown })
+	}
+
+	/// True if the backend has not shutdown
+	pub async fn is_connected(&self) -> bool {
```
Comment (niklasad1) [inline ws-client/src/client.rs:184]: missed this one, doesn't need to be async anymore

```suggestion
	pub fn is_connected(&self) -> bool {
```

File: ws-client/src/client.rs [hunk 6]
```diff
@@ -167,8 +175,14 @@ impl WsClient {
 				max_concurrent_requests,
 			)
 			.await;
+			has_shutdown_sender.store(true, Ordering::Relaxed);
 		});
-		Ok(Self { to_back, request_timeout, error: Mutex::new(ErrorFromBack::Unread(err_rx)) })
+		Ok(Self { to_back, request_timeout, error: Mutex::new(ErrorFromBack::Unread(err_rx)), has_shutdown })
+	}
+
+	/// True if the backend has not shutdown
```
Comment (niklasad1) [inline ws-client/src/client.rs:183]:
```suggestion
	/// Checks if the client is connected to target.
```

File: ws-client/src/client.rs [hunk 7]
```diff
@@ -167,8 +175,14 @@ impl WsClient {
...
 			.await;
+			has_shutdown_sender.store(true, Ordering::Relaxed);
 		});
-		Ok(Self { to_back, request_timeout, error: Mutex::new(ErrorFromBack::Unread(err_rx)) })
+		Ok(Self { to_back, request_timeout, error: Mutex::new(ErrorFromBack::Unread(err_rx)), has_shutdown })
+	}
+
+	/// Checks if the client is connected to the target.
+	pub fn is_connected(&self) -> bool {
+		!self.has_shutdown.load(Ordering::Relaxed)
```
Comment (maciejhirsz) [inline ws-client/src/client.rs:185]: I think (I'm not 100% sure) we could drop the atomic bool entirely and just check if the sender to the background task is still operational here?

```rust
!self.to_back.is_closed()
```

Comment (niklasad1) [reply ws-client/src/client.rs:185]: Sorry that I didn't spotted this, pushed a PR to your branch with a fix.

**Review by maciejhirsz — APPROVED** [review]
Clean :)

**Pull request merged**